hbase-server/src/main/java/org/apache/hadoop/hbase/master/procedure/SnapshotVerifyProcedure.java

```diff
@@ -0,0 +1,206 @@
+          LOG.warn("retry {} on new target server {}", this, s);
+          this.targetServer = s;
+          dispatched = false;
+        });
+        throw new ProcedureYieldException();
+      }
+    } else {
+      return res;
+    }
+  }
+
+  @Override
+  protected void rollback(MasterProcedureEnv env) {
+    // nothing to rollback
+  }
+
+  @Override
+  protected boolean abort(MasterProcedureEnv env) {
+    return false;
+  }
+
+  @Override
+  protected void serializeStateData(ProcedureStateSerializer serializer) throws IOException {
+    SnapshotVerifyProcedureStateData.Builder builder =
+      SnapshotVerifyProcedureStateData.newBuilder();
+    builder.setTargetServer(ProtobufUtil.toServerName(targetServer));
+    builder.setSnapshot(snapshot);
+    for (RegionInfo region : regions) {
+      builder.addRegion(ProtobufUtil.toRegionInfo(region));
+    }
+    builder.setExpectedNumRegion(expectedNumRegion);
+    serializer.serialize(builder.build());
+  }
+
+  @Override
+  protected void deserializeStateData(ProcedureStateSerializer serializer) throws IOException {
+    SnapshotVerifyProcedureStateData data =
+      serializer.deserialize(SnapshotVerifyProcedureStateData.class);
+    this.targetServer = ProtobufUtil.toServerName(data.getTargetServer());
+    this.snapshot = data.getSnapshot();
+    this.regions = data.getRegionList().stream()
+      .map(ProtobufUtil::toRegionInfo).collect(Collectors.toList());
+    this.expectedNumRegion = data.getExpectedNumRegion();
+  }
+
+  @Override
+  public Optional<RemoteProcedureDispatcher.RemoteOperation> remoteCallBuild(
+      MasterProcedureEnv env, ServerName serverName) {
+    SnapshotVerifyParameter.Builder builder = SnapshotVerifyParameter.newBuilder();
+    builder.setSnapshot(snapshot);
+    for (RegionInfo region : regions) {
+      builder.addRegion(ProtobufUtil.toRegionInfo(region));
+    }
+    builder.setExpectedNumRegion(expectedNumRegion);
+    return Optional.of(new RSProcedureDispatcher.ServerOperation(this, getProcId(),
+      SnapshotVerifyCallable.class, builder.build().toByteArray()));
+  }
+
+  @Override
+  public ServerName getServerName() {
+    return targetServer;
+  }
+
+  @Override
+  public boolean hasMetaTableRegion() {
+    return false;
+  }
+
+  @Override
+  public ServerOperationType getServerOperationType() {
+    return ServerOperationType.VERIFY_SNAPSHOT;
+  }
+
+  private Optional<SnapshotProcedure> getParent(MasterProcedureEnv env) {
+    SnapshotProcedure parent = env.getMasterServices().getMasterProcedureExecutor()
+      .getProcedure(SnapshotProcedure.class, getParentProcId());
+    if (parent == null) {
+      LOG.warn("Parent procedure of {} does not exist. Is it created without "
+        + "a SnapshotProcedure as parent?", this);
+      return Optional.empty();
+    } else {
+      return Optional.of(parent);
+    }
+  }
+
+  private Optional<ServerName> newTargetServer(MasterProcedureEnv env) {
+    List<ServerName> onlineServers =
```

Comment:
We have implemented a ReservoirSample class in hbase-common, you can this this class to do random selection, without copying all the online servers out.

Reply:
Ok. I will fix it. Thank you.